File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/require-hook-stage.ts
#!/usr/bin/env bun
import type { Finding, HarnessCheckRule, HarnessManifest, RuleContext } from "../harness-check-rule";

const STACK = "bun" as const;

/**
 * Require hooks to contain stage markers.
 */
export class RequireHookStageRule implements HarnessCheckRule {
  static readonly category = "requireHookStage";

  constructor(private readonly ctx: RuleContext) {}

  applies(manifest: HarnessManifest): boolean {
    const section = manifest.requireHookStage;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    if (enabled === false) {
      return false;
    }
    const parameters = this.ctx.readJsonObject((section as Record<string, unknown>).parameters);
    const stages = this.ctx.readJsonObject(parameters.stages);
    return this.ctx.readJsonObject(stages[STACK]).length > 0;
  }

  validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
    const section = this.ctx.readJsonObject(manifest.requireHookStage);
    const parameters = this.ctx.readJsonObject(section.parameters);
    const hooks = this.ctx.readStringArray(parameters.hooks);
    const markerTemplate = typeof parameters.markerTemplate === "string" ? parameters.markerTemplate : "";
    const stagesEntry = this.ctx.readJsonObject(parameters.stages);
    const stackStages = this.ctx.readJsonObject(stagesEntry[STACK]);

    return hooks.flatMap((hook) => {
      if (!this.ctx.isFile(hook)) {
        return [];
      }
      const hookName = hook.split("/").pop() ?? "";
      const stageKey = hookName === "pre-commit" ? "pre-commit" : "pre-push";
      const stage = typeof stackStages[stageKey] === "string" ? stackStages[stageKey] : "";
      if (!stage) {
        return [];
      }
      const marker = markerTemplate.replace("{stage}", stage);
      const text = this.ctx.read(hook);

      return text.includes(marker)
        ? []
        : [
            {
              severity: this.ctx.severityOf(manifest, RequireHookStageRule.category),
              category: RequireHookStageRule.category,
              message: `${hook} must contain stage marker '${marker}'`,
            },
          ];
    });
  }
}

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/require_hook_stage.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require hook stage rule."""

from collections.abc import Iterable

from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import is_safe_file, read_text, severity_for

STACK = "uv"

class RequireHookStageRule(HarnessCheckRule):
    """Validate requireHookStage check."""

    category = "requireHookStage"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate requireHookStage check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        marker_template = params.get("markerTemplate", "# Harness stage: {stage}")
        stages = params.get("stages", {})
        if not isinstance(stages, dict):
            return []
        stack_stages = stages.get(STACK, {})
        if not isinstance(stack_stages, dict):
            return []
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        template = messages.get("default", "{hook} must contain stage marker '# Harness stage: {expectedStage}'")
        result = []
        for hook_name, expected_stage in stack_stages.items():
            if not isinstance(hook_name, str):
                continue
            if not isinstance(expected_stage, str):
                continue
            path = root / f"docs/harness/git-hooks/{hook_name}"
            if not is_safe_file(path):
                continue
            text = read_text(path)
            if marker_template.format(stage=expected_stage) not in text:
                result.append(Finding(
                    severity_for(manifest, self.category),
                    self.category,
                    template.format(hook=hook_name, expectedStage=expected_stage),
                ))
        return result
